### PR TITLE
feat(errors): improved error handling

### PR DESCRIPTION
**Pull request by jimeh**
Errors are now handled in a multi-step process. They are first unmarshaled to a
`*katapult.ResponseError`, the `Code` field is then checked for a known error
code, and if found, the error is changed to the specific struct for that error.

This allows errors with varying types and fields on the Detail field to be
separated parsed/unmashaled with specific structs.

The error casting process happens at two layers, first within the `katapult`
package itself, and if it can't cast it to a specific error type, it just
returns a `*katapult.ResponseError`.

The `core` package handles errors by attempting to type assert them to a
`*katapult.ResponseError`, and if successful, it performs its own error casting
based on the errors specific to the Core API.

All error structs embed the `katapult.CommonError` type, which ensures they
implement the `error` interface and can be passed around like normal Go errors.
Also this embedding allows the `Error()` method on error structs to be
overridden manually, which is how we return a custom error string for errors
which have various types detail values. This seems sensible as the plain text
formatting/wording is highly dependent on the context of what kind of error it
is, what it's describing, and what kind of details are available.

This is primarily not a breaking a change, as method signatures are the same,
but errors returned from all methods now contain and convey more detailed
information. The only kind of breakage is that some errors now render different
text via the `Error()` method.

This has a lot of nice benefits, previously if to determine if there was a legit
error in fetching a resource, or it simply doesn't exist, you had to do
something like this:

```go
ip, resp, err := client.IPAddresses.GetByID(ctx, "ip_r3NoJOsbEtZOwT89")
if err != nil {
	if resp != nil && resp.Response != nil && resp.StatusCode == 404 {
		// do something for not found errors, like create the resource
	} else {
		// do something for any other error
	}
}
```

While with this change you can simply do this:

```go
ip, _, err := client.IPAddresses.GetByID(ctx, "ip_r3NoJOsbEtZOwT89")
if err != nil {
	if errors.Is(err, katapult.ErrNotFound) {
		// do something for not found errors, like create the resource
	} else {
		// do something for any other error
	}
}
```

And the error chain is quite detailed, so when a IP address is not found for
example, the `error` that is returned is actually a
`core.IPAddressNotFoundError` type. The error chain from that type is:

- `core.ErrIPAddressNotFound`
- `katapult.ErrResourceNotFound`
- `katapult.ErrNotFound`
- `katapult.ErrResponse`
- `katapult.Err`

This makes it easy to check exactly what kind of error it is you're dealing
with, or checking if it's in general a not found error or any kind, etc.

Additionally, you can directly access all the details fields of an error through
type assertions. For example if you get a `scope_not_granted` error, you can
access the string slice of required scopes like this:

```go
if e, ok := err.(*katapult.ScopeNotGrantedError); ok {
	fmt.Printf("%+v\n", e.Detail.Scopes) // => [core.ip core.ip:read]
}
```

### CODE REVIEW

**Comment by jimeh** [comment]
Thinking about this a bit more, I think the specific error types can be
simplified a bit more, and effectively just return `fmt.Errorf()` after the
Detail raw value has been parsed. Though to still be able to customize the
rendering, we'd still need a `Error()` function, but we can rename it to
something else as the struct itself wouldn't need to implement the `error`
interface anymore.

File: core/errors.go [hunk 1]
```diff
@@ -1,7 +1,23 @@
...
 
 import (
-	"errors"
+	"fmt"
+
+	"github.com/krystal/go-katapult"
 )
 
-var Err = errors.New("")
+//go:generate go run github.com/krystal/go-katapult/tools/codegen -t errors -p core -o . -i CoreAPI\/.* -n core/v1
```
Comment (strideynet) [inline core/errors.go:9]: Does this definitely use the local version when specified like this?

Comment (jimeh) [reply core/errors.go:9]: Yep, it does, cause Go is aware that the module name we're within is `github.com/krystal/go-katapult` from the `go.mod` file.

Also, if it didn't work, none of my local generation would have worked so far :)

File: errors.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+// CommonError handles common logic shared between all API-based error types.
+type CommonError struct {
+	parent error
+
+	Code        string `json:"code,omitempty"`
+	Description string `json:"description,omitempty"`
+}
+
+func NewCommonError(parent error, code, description string) *CommonError {
```
Comment (strideynet) [inline errors.go:151]: Perhaps we ought to return a value type here? It seems weird that we have to use indirection in the generated code.

Comment (jimeh) [reply errors.go:151]: Yeah, agreed. I was kind of undecided about it cause it felt wasteful copying the whole struct, but I think only the `Code` and `Description` fields which would be copied as the `parent` field should be a pointer to a `error`, it's probably fine.

**Review by strideynet — APPROVED** [review]

**Review by strideynet — APPROVED** [review]

**Pull request merged**